src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs

```diff
@@ -79,19 +83,53 @@ public static Completion.CompletionTrigger LSPToRoslynCompletionTrigger(LSP.Comp
             }
             else if (context.TriggerKind == LSP.CompletionTriggerKind.Invoked)
             {
-                return Completion.CompletionTrigger.Invoke;
+                if (context is not LSP.VSCompletionContext vsCompletionContext)
+                {
+                    return Completion.CompletionTrigger.Invoke;
+                }
+
+                switch (vsCompletionContext.InvokeKind)
+                {
+                    case LSP.VSCompletionInvokeKind.Explicit:
+                        return Completion.CompletionTrigger.Invoke;
+
+                    case LSP.VSCompletionInvokeKind.Typing:
+                        var insertionChar = await GetInsertionCharacter(document, position, cancellationToken).ConfigureAwait(false);
+                        return Completion.CompletionTrigger.CreateInsertionTrigger(insertionChar);
+
+                    case LSP.VSCompletionInvokeKind.Deletion:
+                        Contract.ThrowIfNull(context.TriggerCharacter);
+                        Contract.ThrowIfFalse(char.TryParse(context.TriggerCharacter, out var triggerChar));
+                        return Completion.CompletionTrigger.CreateDeletionTrigger(triggerChar);
+
+                    default:
+                        // LSP added an InvokeKind that we need to support.
+                        Logger.Log(FunctionId.LSPCompletion_MissingLSPCompletionInvokeKind);
+                        return Completion.CompletionTrigger.Invoke;
+                }
             }
             else if (context.TriggerKind == LSP.CompletionTriggerKind.TriggerCharacter)
             {
                 Contract.ThrowIfNull(context.TriggerCharacter);
-                return Completion.CompletionTrigger.CreateInsertionTrigger(char.Parse(context.TriggerCharacter));
+                Contract.ThrowIfFalse(char.TryParse(context.TriggerCharacter, out var triggerChar));
+                return Completion.CompletionTrigger.CreateInsertionTrigger(triggerChar);
             }
             else
             {
                 // LSP added a TriggerKind that we need to support.
                 Logger.Log(FunctionId.LSPCompletion_MissingLSPCompletionTriggerKind);
                 return Completion.CompletionTrigger.Invoke;
             }
+
+            // Local functions
+            static async Task<char> GetInsertionCharacter(Document document, int position, CancellationToken cancellationToken)
+            {
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                // We use 'position - 1' here since we want to find the character that was just inserted.
+                var triggerCharacter = char.Parse(text.ToString(new TextSpan(position - 1, 1)));
```

Comment:
Seems odd to call ToString just to get one parameter. Is it possible to use [the indexer](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Text/SourceText.cs,445) on `text` or does it not does `position` not map to an index like that?

Reply:
also, how do we know the position > 0 ?